File: Data Structures & Algorithms/partition-equal-subset-sum/submission-12.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {boolean}
     */
    canPartition(nums) { 
        const sum = nums.reduce((acc, cur) => acc + cur, 0) 

        if (sum % 2 !== 0) return false 

        return this.dfs(0, 0, nums, sum / 2, {})
    } 

    dfs(start, cur, nums, target, memo) { 
        const key = `${start}_${cur}` 
        if (key in memo) return memo[key]
        if (cur === target) return true 
        if (cur > target) return false 

        for (let i = start; i < nums.length; i ++) {
            const res = this.dfs(i + 1, cur + nums[i], nums, target, memo)  
            if (res) {
                memo[key] = true 
                return true
            }
        }

        memo[key] = false
        return false
    }
}

File: Data Structures & Algorithms/longest-palindromic-substring/submission-12.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        start = 0 
        length = 0 

        for i in range(len(s)):  
            l = i 
            r = i
            while l >= 0 and r < len(s) and s[l] == s[r]: 
                if r - l + 1 > length: 
                    length = r - l + 1 
                    start = l 

                l -= 1 
                r += 1  

         
        for i in range(len(s)):  
            l = i 
            r = i + 1
            while l >= 0 and r < len(s) and s[l] == s[r]: 
                if r - l + 1 > length: 
                    length = r - l + 1 
                    start = l 

                l -= 1 
                r += 1 

        return s[start: start + length]

        